src/box/txn.c

```diff
@@ -1497,10 +1523,16 @@ txn_on_yield(struct trigger *trigger, void *event)
 	struct txn *txn = in_txn();
 	assert(txn != NULL);
 	if (txn->status != TXN_ABORTED && !txn_has_flag(txn, TXN_CAN_YIELD)) {
-		txn_rollback_to_svp(txn, NULL);
+		txn_rollback_to_svp(txn, NULL, true);
 		txn->status = TXN_ABORTED;
 		txn_set_flags(txn, TXN_IS_ABORTED_BY_YIELD);
 		say_warn("Transaction has been aborted by a fiber yield");
+		if (txn_has_flag(txn, TXN_HAS_TRIGGERS)) {
+			/* Can't rollback more than once. */
+			trigger_destroy(&txn->on_rollback);
+			/* Commit won't happen after rollback. */
+			trigger_destroy(&txn->on_commit);
+		}
```

Comment:
From that moment you have the destroyed triggers here but the flag `TXN_HAS_TRIGGERS` is still set. Either the triggers need to be re-created empty, or the flag needs to be unset.

Reply:
AFAIU after `trigger_destroy` is called, the trigger list becomes empty, but yeah, it's probably better to reinitialize it explicitly.